### PR TITLE
[Logs]: Isolated the logic to seek in a dedicated class and helpers

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

Factorise and isolated  the logic to compute the read offset for tailers:
- Created a dedicated package `pkg/logs/seek`
- Inject a new `Seeker` in all scanner/launcher which needed the auditor before
- Created dedicated helper for `file` and `docker`

### Motivation

Configs are not available at agent start anymore which means we can't use the previous logic to know where to start tailing elements from.
The logic is now based on `ctime` or creation time:
- if the element has already been tailed in the past -> tail from offset from registry
- if the element is created before the agent start -> tail from end
- if the element is created after the agent start -> tail from beginning

### Additional Notes

Still need to write unit-tests

### CODE REVIEW

File: pkg/logs/seek/seek.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+// NewSeeker returns a new seeker
+func NewSeeker(registry Registry) *Seeker {
+	return &Seeker{
+		registry: registry,
+	}
+}
+
+// origin represents the time reference of the agent.
+var origin = time.Now()
```
Comment (mnshdw) [inline pkg/logs/seek/seek.go]: Let's consider this scenario:
- agent is started
- daylight saving time occurs, shifting current time by -1h
- a new file is created, and tailed by the agent

Are we going to consider the file was created before the agent start, thus tail from the beginning ?

Not sure it's a problem, I'm just trying thinking of the possible pitfalls of storing the start time this way.

File: pkg/logs/input/docker/scanner.go [hunk 2]
```diff
@@ -144,21 +144,25 @@ func (s *Scanner) setup() error {
...
 // returns true if the setup succeeded, false otherwise
-func (s *Scanner) setupTailer(cli *client.Client, container types.Container, source *config.LogSource, tailFromBeginning bool, outputChan chan message.Message) bool {
+func (s *Scanner) setupTailer(cli *client.Client, container types.Container, source *config.LogSource, outputChan chan message.Message) bool {
 	log.Info("Detected container ", container.Image, " - ", s.humanReadableContainerID(container.ID))
-	t := NewTailer(cli, container.ID, source, outputChan)
-	err := t.recoverTailing(s.auditor, tailFromBeginning)
+	tailer := NewTailer(cli, container.ID, source, outputChan)
+	since, err := Since(s.seeker, container, tailer.Identifier())
+	if err != nil {
+		log.Warnf("Could not recover last committed offset for container %v: %v", container.ID[:12], err)
```
Comment (mnshdw) [inline pkg/logs/input/docker/scanner.go]: I'd be wary of static `:12` a bit everywhere in the code which is very docker-tainted, we might stumble one day on a container runtime that has shorter ids. I think we can spare a short method somewhere that checks the length of the id before returning the shortened version.

**Review by mnshdw — DISMISSED** [review]
I like the change, good simplification. I left a few minor comments.

It's hard to assess whether there is any regression just looking at the code so I'm going to trust your unit tests :p

File: pkg/logs/input/docker/scanner.go [hunk 3]
```diff
@@ -144,21 +144,25 @@ func (s *Scanner) setup() error {
...
 // setupTailer sets one tailer, making it tail from the beginning or the end,
 // returns true if the setup succeeded, false otherwise
-func (s *Scanner) setupTailer(cli *client.Client, container types.Container, source *config.LogSource, tailFromBeginning bool, outputChan chan message.Message) bool {
-	log.Info("Detected container ", container.Image, " - ", s.humanReadableContainerID(container.ID))
-	t := NewTailer(cli, container.ID, source, outputChan)
-	err := t.recoverTailing(s.auditor, tailFromBeginning)
+func (s *Scanner) setupTailer(cli *client.Client, container types.Container, source *config.LogSource, isSetup bool, outputChan chan message.Message) bool {
+	log.Info("Detected container ", container.Image, " - ", ShortContainerID(container.ID))
+	tailer := NewTailer(cli, container.ID, source, outputChan)
+	since, err := Since(s.registry, tailer.Identifier(), isSetup)
```
Comment (tmichelet) [inline pkg/logs/input/docker/scanner.go]: `isSetup` is confusing, we don't know if it is about the tailer or the agent, or the scanner

File: pkg/logs/input/file/position.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	var offset int64
+	var whence int
+	var err error
+	value := registry.GetOffset(identifier)
+	if value != "" {
+		offset, err = strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			offset, whence = 0, io.SeekEnd
+		} else {
+			whence = io.SeekStart
```
Comment (tmichelet) [inline pkg/logs/input/file/position.go]: why setting the offset elsewhere but not here?

Comment (tmichelet) [reply pkg/logs/input/file/position.go]: ah nevermind

Comment (tmichelet) [reply pkg/logs/input/file/position.go]: I think to clarify, I'd put `whence = io.SeekStart` before the `offset, err = strconv.ParseInt(value, 10, 64)`
I'm not fan of `if err == nil set it` I find it harder to read

**Review by tmichelet — APPROVED** [review]

**Pull request merged**